tests/e2e/hashkv_test.go

```diff
@@ -0,0 +1,314 @@
+			OnlyOneKey:     false,
+		},
+	}
+
+	for idx, tt := range tests {
+		for _, scenario := range scenarios {
+			t.Run(fmt.Sprintf("#%d - %s - OnlyOneKey=%v", idx, scenario.ClusterVersion, scenario.OnlyOneKey), func(t *testing.T) {
+				ctx := context.Background()
+
+				clus, cliCfg := newClusterForHashKV(t, scenario.ClusterVersion)
+
+				newTestKeySetInCluster(t, clus, cliCfg, scenario.OnlyOneKey)
+
+				cli, err := fe2e.NewEtcdctl(cliCfg, clus.EndpointsGRPC())
+				require.NoError(t, err)
+
+				t.Logf("COMPACT rev=%d", tt.compactedOnRev)
+				_, err = cli.Compact(ctx, tt.compactedOnRev, config.CompactOption{Physical: true})
+				require.NoError(t, err)
+
+				t.Logf("HashKV on rev=%d", tt.hashKVOnRev)
+				resp, err := cli.HashKV(ctx, tt.hashKVOnRev)
+				require.NoError(t, err)
+
+				require.Len(t, resp, 3)
+				require.True(t, resp[0].Hash != 0)
+				t.Logf("One Hash value is %d", resp[0].Hash)
+
+				require.Equal(t, resp[0].Hash, resp[1].Hash)
+				require.Equal(t, resp[1].Hash, resp[2].Hash)
+			})
+		}
+	}
+}
+
+// key: "foo"
+// modified: 41
+// generations:
+//
+//	{34, 0}, {35, 0}, {36, 0}, {37, 0}, {38, 0}, {39, 0}, {40, 0}, {41, 0}
+//	{23, 0}, {24, 0}, {25, 0}, {26, 0}, {27, 0}, {28, 0}, {29, 0}, {30, 0}, {31, 0}, {32, 0}, {33, 0}[tombstone]
+//	{12, 0}, {13, 0}, {14, 0}, {15, 0}, {16, 0}, {17, 0}, {18, 0}, {19, 0}, {20, 0}, {21, 0}, {22, 0}[tombstone]
+//	{2, 0},  {3, 0},  {4, 0},  {5, 0},  {6, 0},  {7, 0},  {8, 0},  {9, 0},  {10, 0}, {11, 0}[tombstone]
+//
+// # If onlyOneKey is false
+//
+// key: "foo-1"
+// modified: 12
+// generations:
+//
+//	{12, 1}
+//
+// key: "foo-2"
+// modified: 23
+// generations:
+//
+//	{23, 1}
+//
+// key: "foo-3"
+// modified: 34
+// generations:
+//
+//	{34, 1}
+func newTestKeySetInCluster(t *testing.T, clus *fe2e.EtcdProcessCluster, cliCfg fe2e.ClientConfig, onlyOneKey bool) {
+	t.Helper()
+	c := newClient(t, clus.EndpointsGRPC(), cliCfg)
+	defer c.Close()
+
+	ctx := context.Background()
+	key := "foo"
+	totalRev := 41
+
+	deleteOnRev := 11 // 22, 33
+	lastRevision := int64(1)
+	for i := 2; i <= totalRev; i++ {
+		require.Equal(t, int64(i-1), lastRevision)
+
+		if i%deleteOnRev == 0 {
+			t.Logf("DELETEing key=%s", key)
+			resp, derr := c.Delete(ctx, key)
+			require.NoError(t, derr)
+			lastRevision = resp.Header.Revision
+			continue
+		}
+
+		value := fmt.Sprintf("%d", i)
+		ops := []clientv3.Op{clientv3.OpPut(key, value)}
+
+		logMsg := fmt.Sprintf("PUTing key=%s", key)
+		if i%deleteOnRev == 1 && !onlyOneKey {
+			key2 := fmt.Sprintf("%s-%d", key, i/deleteOnRev)
+			ops = append(ops, clientv3.OpPut(key2, value))
+			logMsg = fmt.Sprintf("%s,key=%s", logMsg, key2)
+		}
+		t.Logf("%s val=%s", logMsg, value)
+
+		resp, terr := c.Txn(ctx).Then(ops...).Commit()
+		require.NoError(t, terr)
+		require.True(t, resp.Succeeded)
+		require.Len(t, resp.Responses, len(ops))
+		lastRevision = resp.Header.Revision
+	}
+}
+
+func TestVerifyHashKVAfterTwoCompactions_MixVersions(t *testing.T) {
```

Comment:
Side note: If https://github.com/etcd-io/etcd/pull/18274 is accepted, the main branch always keep compacted revision. However, existing releases delete that compacted revisions if they're tombstone.  The #18274 should skip previous compacted revision which is tombstone. So that HashKV result can be the same in mix versions, especially it won't file data corrupt alert when cluster is updating to new release.